### ISSUE TITLE
Investigate real in-place patching (without using a staging folder)

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/butler#118]
So once upon a time, it was very desirable to write all "touched" files to disk in a "-staging" folder because healing was not a thing yet. So if patching failed we definitely did not want to end up with a corrupted game folder, yes?

But now that healing is a thing - if patching fails, who cares! We can just redownload the bits we care about, fresh off the CDN.

So that means we can maybe be more aggressive when patching, e.g. writing directly to the target file and keeping the bits we need in RAM - if they fit.

---

Now it's something we need to carefully think about before implementing because we don't want corrupted patching to happen just because the algorithm is wrong, but basically

  * We need to know what ranges of the old file we're using to make the new file (we know that)
  * We need to check before we write to the new file (which is the same as the old file now), that we're not going to need that bit later on when applying the patch
  * If we are going to need that bit later on when applying the patch, we should try to allocate memory for that bit & save it - if that fails, we should write it to disk.

---

So let's take an example then, if we go from:

`ABBBBBBBBBBBB`

to

`BBBBBBBBBBBBA`

Then when we start writing the `B` in the first position, we need to keep the `A` somewhere so that we can write it when we end up patching the final position.

---

Now an interesting way to look at this is to go the opposite way and, instead of thinking of it as "going through the new file, figuring out what to write op by op, taking from the old file(s) as we go", we might think of it as "for each re-used block, let's find out where it goes in the new file - and if it overwrites an old block that's re-used elsewhere, let's save that in RAM".

That particular strategy could still fail with lack of RAM, but I think that'd be a lot harder (to make it fail). Because then at each step, if we *just saved* blocks, we'll try to put them where they belong now (and save any more if needed) - and that shouldn't explode.

If you're reading this and it doesn't make any sense, don't worry, I'm sorry I can't write it in a clearer way right now.

I'm not entirely sure that strategy is doable for bsdiff patches, although at first sight, I don't see why it wouldn't. The thing with bsdiff patch is that the instructions are not aligned with blocks - so we can't work with blocks, only ranges. And that complicates things because instead of having, say, an array of bools, we have to have sorted pairs of indices, and do a bunch of comparisons there (or maybe they could be trees instead, for faster lookups. ah well)

**Comment by fasterthanlime** [comment]
Tentative explanation of the `inverse apply` strategy, say the old file is:

`A000B111C222`

And the new file is:

`C000A111B222`

---

The ops are: re-use C, re-use 000, re-use A, re-use 000, re-use B, re-use 000.

Now when applying, we'll take the first re-use operation (C) and figure out where it goes. We see that it overwrites `A`, which is also re-used later, so we save `A` somewhere - now we have:

`C000B111C222` + `A`

Then we try to see where `A` is re-used, and it's a single place, and it overwrites B, so we put A where it belongs and save `B` for later - now we have:

`C000A111C222` + `B`

Then we try to see where `C` is re-used, and it's a single place, and `C` has already been re-used, so we just put it where it needs to be:

`C000A111B222`

And - voilà! In-place patching.

**Comment by fasterthanlime** [comment]
Of course, the main issue with that (apart from: it's complicated, so it'll be hard to get right) is that patches are a stream of operations - there's no index that tells us (see previous comment) if and where `A` will be re-used when we want to overwrite it.

We don't want to lose the "stream of ops" thing though, it's very nice when pushing a patch, and it's very nice when applying a patch while downloading it too. So maybe we need a new subtype, a "re-use map", that tells us exactly what is re-used and where. And now it's starting to be a bit complicated :)

But I still think it's worth looking into!

**Comment by fasterthanlime** [comment]
Before going all out and thinking about "re-use map" files though, it'd be worth trying to see what the RAM usage would look like for the naive method (which is to save something as soon as it's overwritten with something else). The big problem is that we might end up saving a bunch of stuff that never gets re-used, for example if we go from 

`AAAAAAAAA`

to

`BBBBBBBBB`

then we'll save all the As in case they get re-used later (because we have no "re-use map") and then we'll realize we never needed them at all.

---

Or maybe there's something smart to be done. We have the full signature of the new build when applying (even when doing streaming apply) - so when we overwrite a block, we can look it up & see if it's in the signature of the new build. If it's not, then there's no need to retain it.

**Comment by fasterthanlime** [comment]
> Or maybe there's something smart to be done. We have the full signature of the new build when applying (even when doing streaming apply) - so when we overwrite a block, we can look it up & see if it's in the signature of the new build. If it's not, then there's no need to retain it.

But wait - one might interject - doesn't that mean we have to hash all the blocks that are overwritten?

Yes! It does mean that. But keep in mind that hashing is pretty fast (here's saints row 3):

```
$ butler sign Saints\ Row\ the\ Third/ /tmp/sr3.pws
∙ Creating signature for Saints Row the Third/
√ 9.3 GiB (234 files, 5 dirs, 0 symlinks) @ 209 MiB/s
```

And it saves a bunch of disk writes (especially if the files involved vary by a small %, which is not terribly uncommon). And disks writes are slow. So I think not only it might be a win in "disk space needed to patch", it might also be a win in "patch speed".

**Comment by fasterthanlime** [comment]
So, this was a fun thread, but it all ignores the fact that "patching" is a regular download operation, which should be pausable/cancellable.

Also, the process that's patching can get killed at any point - OOM killer, machine shutting down, power failure, act of dog, you name it.

So writing directly to the target file is way too risky. The way I ended up implementing this was by rewriting the patching logic to work with "bowls":

  * A "fresh bowl" applies a patch from a folder to a completely different folder (we don't have to worry about overwriting old files!)
  * An "overlay bowl" uses a staging folder where it stores simple "patches" that contain only changed byte ranges

After patching, bowls are "committed". For fresh, that's a no-op. For overlay, the changed byte ranges are copied to the new file.

(Note: `overlay` is its own package: https://github.com/itchio/wharf/tree/master/pwr/overlay - it's the simplest form of patching you can imagine).

This has nice properties, for example:

  * Patch application can be paused & resumed from the last checkpoint
  * Checkpointing required saving the brotli decoder state, which we did in https://github.com/itchio/savior
  * Even if we get interrupted in the middle of "committing" a bowl, committing it again doesn't result in corrupted output:
    * In fresh's case, it no-ops twice
    * In overlay's case, an overlay patch can be applied any number of times with the same result, as it doesn't re-use any of the old data (it's just a collection of fresh byte ranges). It doesn't matter where it gets interrupted as long as it eventually completes successfully.

Incidentally, this means we only have to mark a cave as `morphing` (cf. https://github.com/itchio/butler/blob/a55b533a17a0fd5fa565b67498bc3f93de25bcce/database/models/cave.go#L24 ) when we're comitting the bowl - since up to that point, nothing has been changed in the install folder yet.

---

In summary: I'm very happy with this solution, as it ended up being relatively simple and elegant - each part is tested individually and I think this time I avoided the over-engineering trap.
